### PR TITLE
Additional checks on PR via GitHub actions

### PR DESCRIPTION
**Pull request by albestro**
After realizing there were files without license header in our codebase, let's see if we can add an action that checks that for us.

- [x] add check for file with non-standard extensions
- [x] add check presence of license header in source files
  - we may want to add a "exclude-list", e.g. [`include/dlaf/cusolver/hegst.h`](https://github.com/eth-cscs/DLA-Future/blob/master/include/dlaf/cusolver/hegst.h)
  - some files have multiple licenses, e.g. all `gtest_*_main` like [test/src/gtest_mpipika_main.cpp](https://github.com/eth-cscs/DLA-Future/blob/master/test/src/gtest_mpipika_main.cpp)
- [x] add check presence of license header in CMake files
  - there is at least a Find*.cmake module which is not checked
- [x] add check for include guards (e.g. pragma once)

### CODE REVIEW

File: .github/workflows/check_license_header.yaml [hunk 1]
```diff
@@ -0,0 +1,79 @@
+name: Check license
+
+on: [pull_request]
+
+jobs:
+  check-cpp:
```
Comment (albestro) [inline .github/workflows/check_license_header.yaml:6]: note to self: change job name

**Review by msimberg — COMMENTED** [review]
This is really nice!

Regarding the files with non-ETHZ licenses, I see two alternatives:
1. Simply ignore them when they fail, but this would require the checks to only work on _changed_ files because otherwise the checks would fail all the time.
2. Allow skipping the check per-file with some magic phrase. In pika there's `pika-no-inspect` or `pika-no-inspect: check-name` for this purpose.

File: .github/workflows/check_license_header.yaml [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+          # Generate header license CMake compliant
+          sed 's|^//|#|g' misc/HEADER > HEADER_CMAKE
+
+          # Compare first lines of cmake files with the reference
+          find  '(' -type f ! -path './misc*' ')'                                 \
+                '('                                                               \
+                        -name 'CMakeLists.txt'                                    \
+                    -o  -name 'DLAF*.cmake'                                       \
+                    -o  -name 'DLAF*.cmake.in'                                    \
+                ')'                                                               \
```
[the remark this answers is not in the document]
Comment (albestro) [reply .github/workflows/check_license_header.yaml:69]: Good point. In case, the check for them is slightly different because they may (or may not ) have the shebang at first line. I'm going to add this check somehow.

About the extension check, I'm not really sure I got the point. Currently what the ext check does is to look inside `src`/`include`/`test` folders for files that have an extension different from ones listed there. It is a bit the "other way around", because it's not checking that C++ files have valid extension, but that inside there there are just files with valid extensions.

Anyway, we could decide to change it to rely on `file` output (iirc it does not rely on the extension), so that any file that is considered a C/C++ file, should have the extensions we want.

And at that point, we could also add the check for python files having `*.py`. However I'm not sure if this is what you have in mind.

**Comment by albestro** [comment]
> Regarding the files with non-ETHZ licenses, I see two alternatives:
> 
> 1. Simply ignore them when they fail, but this would require the checks to only work on _changed_ files because otherwise the checks would fail all the time.

I have to think about it a moment, but it should not be a super difficult script upgrade.

> 2. Allow skipping the check per-file with some magic phrase. In pika there's `pika-no-inspect` or `pika-no-inspect: check-name` for this purpose.

This is something that also @rasolca mentioned to me.


What would you prefer @rasolca @teonnik?

File: .github/workflows/check_license_header.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (msimberg) [reply .github/workflows/check_license_header.yaml:69]: > About the extension check, I'm not really sure I got the point. Currently what the ext check does is to look inside `src`/`include`/`test` folders for files that have an extension different from ones listed there. It is a bit the "other way around", because it's not checking that C++ files have valid extension, but that inside there there are just files with valid extensions.
> 
> Anyway, we could decide to change it to rely on `file` output (iirc it does not rely on the extension), so that any file that is considered a C/C++ file, should have the extensions we want.
> 
> And at that point, we could also add the check for python files having `*.py`. However I'm not sure if this is what you have in mind.

Yeah, I didn't think clearly with the comment about the extensions. Somehow I was first thinking that you were looking to have a global whitelist-type of thing for allowed extensions in the repo, but it makes a lot less sense for python where the only reasonable extension is `.py`. For C++ it makes more sense to check the extensions since there are a lot of options, and I guess this was your main motivation for adding that check. So I think you can ignore my comment about the extensions...

[the remark this answers is not in the document]
Comment (albestro) [reply .github/workflows/check_license_header.yaml:69]: > For C++ it makes more sense to check the extensions since there are a lot of options, and I guess this was your main motivation for adding that check

Exactly! As I said it is a bit the other way around, because it is not

`C++ file => check extension is in the list`

but it is

`a file in src/include/test must be either a C++ or CMake file => check extension is in the list`

The alternative is to rely on `file` output, with something like

`if file tool says it is a c/c++ file => check extension is in the list`

[the remark this answers is not in the document]
Comment (msimberg) [reply .github/workflows/check_license_header.yaml:69]: In my opinion what you have right now is reasonable and quite clear.

**Comment by rasolca** [comment]
1. It can work. However it would not help when updating the license unless all the files are scanned when header change. Moreover it doesn't check multiple license headers.

2. This solution in more versatile.

File: .github/workflows/check_license_header.yaml [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            | xargs -I{} sh -c                                                    \
+                "head -n9 {} | diff -sq - HEADER_HASH > /dev/null || echo {}"     \
+            >> result.check
+
+          # Look for the LICENSE inside the python file
+          function streambin { xxd -p $1 | tr -d "\n"; }
+          BIN_HEADER_CMAKE=`streambin HEADER_HASH`
+          for file in `find scripts -type f -name "*.py"`; do                     \
+            streambin $file | grep -sq $BIN_HEADER_CMAKE || echo $file;           \
+          done >> result.check
```
Comment (albestro) [inline .github/workflows/check_license_header.yaml:79]: I wasn't able to figure out an easier and more neat solution.

Since grep does not deal with multilines, the workaround is to see script files as "binary" and then look for the exact ASCII code sequence representing the LICENSE.

File: .github/workflows/check_license_header.yaml [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+          find src include test miniapp -type f                                   \
+               '('                                                                \
+                       -name '*.cpp'                                              \
+                   -o  -name '*.h'                                                \
+                   -o  -name '*.h.in'                                             \
+                   -o  -name '*.tpp'                                              \
+                   -o  -name '*.cu'                                               \
+               ')'                                                                \
+            | xargs -I{} sh -c                                                    \
+                "grep -sq 'dlaf-no-license-check' {}                              \
```
Comment (albestro) [inline .github/workflows/check_license_header.yaml:47]: This is where the special tag check is done. It is valid just for source code files.

File: .github/check_license.sh [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+# SPECIAL TAGs
+
+# note: skip the check
+grep -sq "dlaf-no-license-check" $FILE_TO_CHECK && exit 0
+
+# note: multi-license check
+if grep -sq "dlaf-multi-license-check" $FILE_TO_CHECK; then
+  MAX_LINE=`cat $FILE_TO_CHECK | wc -l | tr -d ' '`
+fi
```
Comment (albestro) [inline .github/check_license.sh:51]: About special tags: it just looks for the tag, no matter where it is (comment, string, ...). Is it good enough?

File: .github/check_license.sh [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  echo "$0 usage: [-r n] license file"
+  echo ""
+  echo "where:"
+  echo "-r n  max line number (1-based) where the first line of license header should start."
+  echo "      First line of the header must fall within range [1,n] (default: 1)"
+  echo ""
+  echo "SPECIAL TAGS: they can be used inside files to control behavior of this tool"
+  echo ""
+  echo "dlaf-no-check-license       skip any check and return 0"
+  echo "dlaf-multi-license-check    check license in the entire file (-r option is ignored)"
```
Comment (albestro) [inline .github/check_license.sh:13]: Multi license check just ignores the range parameter, i.e. if a file has the tag in it, the license can be wherever in it.

File: .github/workflows/check_license_header.yaml [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        run: |
+          # Generate header license CMake/Python compliant (with hashes)
+          sed 's|^//|#|g' misc/HEADER > HEADER_HASH
+
+          # Check license in CMake files
+          find  '(' -type f ! -path './misc*' ')'                                 \
+                '('                                                               \
+                        -name 'CMakeLists.txt'                                    \
+                    -o  -name 'DLAF*.cmake'                                       \
+                    -o  -name 'DLAF*.cmake.in'                                    \
```
[the remark this answers is not in the document]
Comment (albestro) [reply .github/workflows/check_license_header.yaml:68]: I don't see any special reason for FindCMake, I just skipped them previously when we didn't have the special tags yet.

About others, just for reference a list grouped by file extensions.

```
# *.sh
./ci/ctest_to_gitlab.sh
./ci/set_github_status.sh
./scripts/build.sh
./.github/check_license.sh
./.github/format.sh
```

```
# *.yaml
./ci/docker/gpu-debug.yaml
./ci/docker/gpu-release.yaml
./ci/docker/cpu-debug.yaml
./ci/docker/cpu-release.yaml
./.github/workflows/check_license_header.yaml
./spack/repo.yaml
```
For both shell and yaml, I don't see any obstacle.

```
# *.py
./scripts/miniapps.py
./scripts/plot_hegst_weak.py
./scripts/plot_hegst_strong.py
./scripts/systems.py
./scripts/plot_chol_weak.py
./scripts/plot_trsm_strong.py
./scripts/postprocess.py
./scripts/gen_weak.py
./scripts/plot_chol_strong.py
./scripts/plot_trsm_weak.py
./scripts/gen_strong.py
./spack/packages/dla-future/package.py
```

About python scripts, we already have the check for the ones inside `scripts` folder. The other one is the spack package, which has the spack license, but I don't know if we are allowed to add ETHZ one. Anyone has any info on this side? However for the moment I think that since it is not already public in the official spack repo, we can do whatever we want with this file (adding a special tag, using just our license, ...).

TL;DR We can add the check for all of them (they all use hash for comments, so there are no big changes to do) and eventually skip it for specific files.

[the remark this answers is not in the document]
Comment (msimberg) [reply .github/workflows/check_license_header.yaml:68]: If it's straightforward enough to add I think it makes sense to check all of them. No knowledge about the spack package but I'd imagine it should possibly have both. Easiest might be to keep the LLNL one and skip the check.

File: .github/workflows/check_license_header.yaml [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+            | xargs -I{} sh -c                                                    \
+                ".github/check_license.sh misc/HEADER {} > /dev/null || echo {}"  \
+            > result-cpp.check
+
+          # Generate an error message for each offending file
+          for filepath in `cat result-cpp.check`; do                              \
+            echo "::error file=$filepath,line=1::check the license in $filepath"; \
+          done
+
+          test ! -s result-cpp.check
```
Comment (teonnik) [inline .github/workflows/check_license_header.yaml:55]: LGTM!

If you would like to avoid the temporary files `<sth>.check` and to avoid spawning a child-process when using `cat`, perhaps you can do something like:

```
<long_find_and_xargs_command> | \
while read filepath; do
    echo $filepath
done
```

Also some refactoring (into a helper script) on the error message and the find-xargs combo can be made if you would like to shorten the code a bit, but this is a non-blocker and can be done in a future PR.

**Review by teonnik — APPROVED** [review]

File: .github/workflows/check_license_header.yaml (hunk 8 above)
Comment (albestro) [reply .github/workflows/check_license_header.yaml:55]: That's a good point, but it is just for the cpp-like check because there is just one "contributor" to `result-cpp.check`, right? Because, for the #-based check the contributors are cmake and python files.

Do you have any suggestion also for them?

Comment (albestro) [reply .github/workflows/check_license_header.yaml:55]: There's another point. After generating error messages within the `for`, there is also to return an error code based on the same input we pipeline to the `for`.

I don't know if it fits somehow in your proposal.

File: .github/workflows/check_license_header.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (albestro) [reply .github/workflows/check_license_header.yaml]: - Error: check the license in ./cmake/FindLAPACK.cmake
it comes from CMake Recipes (ETHZ), we can just add `dlaf-no-check-license`
- Error: check the license in ./cmake/FindMKL.cmake
it comes from CMake Recipes (ETHZ), we can just add `dlaf-no-check-license`

About next twos, as already said, it is a transitional setup, until we will release it in the official spack repo, when we can drop the `repo.yaml`, and `dla-future/package.py` will probably require to have just a single spack-related license.

Until then, we can decide to have 
- Error: check the license in ./spack/repo.yaml
We can simply add our license, even if it is a minor file.
- Error: check the license in ./spack/packages/dla-future/package.py
We can have it multi-license for the moment.

Waiting for your feedback @msimberg @rasolca @teonnik before these last fixes.

[the remark this answers is not in the document]
Comment (msimberg) [reply .github/workflows/check_license_header.yaml]: I don't think this is a blocker but you could have a "minimal license check" which only checks that the file contains `SPDX-License-Identifier:` and would at least somewhat cover the corner cases. But I think what you've suggested in the above comment is also reasonable.

**Comment by albestro** [comment]
bors try
